added upload

cseed opened the pull request:
upload is a service that displays uploaded data.  It uses Google authentication to only allow the Hail team to view uploads.  Pipeline uploads can be enabled with `enable_pipeline_upload(email)` and you can upload the current Hail log with `hl.upload_log(email)`.  Pipeline uploads happen in the background asynchronously (and errors are ignored).  Log upload is synchronous.

I generated certs for upload and this is the upload that is currently deployed.

cseed commented:
@konradjk Turn on, please.  Example:


```
import hail as hl

email = 'cseed@broadinstitute.org'

hl.init()
hl.enable_pipeline_upload(email)

mt = hl.import_vcf('sample.vcf')
mt = mt.filter_rows(hl.len(mt.alleles) == 2)
print(mt.write('sample.mt', overwrite=True))

hl.upload_log(email)
```

konradjk commented:
Love it! Though email, not slack integration? 😛

tpoterba commented on hail/src/main/scala/is/hail/expr/ir/Interpret.scala:
```diff
@@ -47,12 +49,31 @@ object Interpret {
+      w.toString
+    }
+
+    if (HailContext.get.uploadEmail != null) {
```
using an option type would make this nicer

danking commented on upload/upload/main.py:
```diff
@@ -0,0 +1,202 @@
+
+def get_flow(state=None):
+  flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+    CLIENT_SECRET, scopes=SCOPES, state=state)
```
so this is using the disk to store some information for the duration of the auth flow? or for a longer duration?

cseed replied:
The file is the oauth2 client secret you get from Google when you set up the client.  It is a mounted (read only) secret.

danking commented on upload/upload/main.py:
```diff
@@ -0,0 +1,202 @@
+def oauth2callback():
+  authorization_response = request.url
+  # FIXME as with url_for(..., _scheme='https')
+  authorization_response = re.sub('^http://', 'https://', authorization_response)
```
this is due to nginx stripping HTTPS, yeah?

cseed replied:
Yes, related to the _scheme='https' stuff above.

danking commented on upload/upload/main.py:
```diff
@@ -0,0 +1,202 @@
+@app.route('/', methods=['GET'])
+def index():
+    if 'email' not in session:
+        return redirect(url_for('login'))
```
Ah, and this means the user logged in successfully modulo bugs in `upload` itself, because the `session` is signed by the secret flask key

cseed replied:
Yes.

danking commented on upload/upload/main.py:
```diff
@@ -0,0 +1,202 @@
+
+@app.route('/logout', methods=['GET'])
+def logout():
+  # FIXME revoke
```
as in make their oauth token invalid? I sort of feel like if we don't do it now it will lay here as a FIXME for a while.

danking commented on upload/upload/gen-secret-key.py:
```diff
@@ -0,0 +1,10 @@
+import os
```
hmm. I think we could actually include secret creation in the yaml file.

Secrets look like:
```
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
data:
  username: YWRtaW4=
  password: MWYyZDFlMmU2N2Rm
```

where the data fields are base64 encoded.

`redeploy` just calls `gen-secret-key.py` to make the secret and slams it into the yaml.

It means the secret changes with each deploy :/ but we have less latent state to manage :).

cseed replied:
So if you want to share a session login between multiple services in the Hail "suite", then we need a single common secret key that they all use.

cseed replied:
This would also invalidate sessions between redeploys which seems wrong.

danking replied:
A common secret key seems achievable. Agreed redeploys in this setting invalidate the session. Not the worst thing since we're talking with a new server? At some point we will have to address these issues to get redeployable k8s.

cseed replied:
There might be some confusion.  I ran this code once and added the generated secret key as a k8s secret.  There is no session invalidation problem.  I just checked it in here for reference in case we need to generate generate another one (e.g. rebuild k8s).

danking commented on site/run-letsencrypt.nginx.conf:
```diff
@@ -2,5 +2,5 @@ server {
     listen [::]:80 default_server;
     root /var/www/html;
-    server_name hail.is www.hail.is ci.hail.is ci2.hail.is scorecard.hail.is test.hail.is;
+    server_name hail.is www.hail.is ci.hail.is scorecard.hail.is upload.hail.is test.hail.is dev1.hail.is;
```
who is dev1?

cseed replied:
Reserved for future use.  It's annoying to get new certs if you want to play with an idea.  Another alternative is to have dev-tools.hail.is/scorecard, etc.

danking commented on hail/src/main/scala/is/hail/Uploader.scala:
```diff
@@ -0,0 +1,139 @@
+    queue.offer((typ, contents, email))
+  }
+
+  def join() {
```
this isn't used yet, right?

cseed replied:
Correct.  I debated calling it in `stop()`, but I prefer to think of the enqueued logging as having no cost and not completely reliable.

danking commented on hail/src/main/scala/is/hail/Uploader.scala:
```diff
@@ -0,0 +1,139 @@
+          .toList),
+      "jvm_runtime" -> JObject(
+        "available_processors" -> JInt(runtime.availableProcessors()),
+        "free_memory" -> JInt(runtime.freeMemory()),
```
free memory will change over time, but this seems to get fixed the first time we upload something?

danking replied:
I'm not sure this will be particularly helpful either way since it's the memory after the pipeline has been run.

cseed replied:
Fair, I can delete this.

cseed commented on upload/upload/main.py:
```diff
@@ -0,0 +1,202 @@
+    CLIENT_SECRET, scopes=SCOPES, state=state)
+
+  # FIXME _scheme='https'
+  flow.redirect_uri = url_for('oauth2callback', _external=True, _scheme='https')
```
OK, so nginx strips off https and I set the header X-Forwarded-Proto: https, but flask (I think this is a bug) doesn't use it in url_for.  This workaround isn't wrong, just annoying, but I don't quite know what the "right" fix is.  There are some stackoverflow answers involving creating Middleware to handle this that I didn't understand/dig into.  I'll remove the FIXMEs and make a comment to this effect.

danking replied:
The other `url_for` calls are hitting redirects in nginx, right? Should we just write our own `url_for` that defaults to `_scheme='https'` and forwards the rest of the kwargs to `flask.url_for`

cseed commented:
I think I addressed all the comments.

I also noticed some get-deployed-sha.sh scripts were broken  (name of deployment was wrong) and I fixed them.

cseed commented:
Also added upload to projects and added upload get-deployed-sha and deploy scripts.

danking commented on batch/get-deployed-sha.sh:
```diff
@@ -1,4 +1,4 @@
 set -ex
 
-kubectl get --selector=app=batch deployments -o "jsonpath={.items[*].metadata.labels.hail\.is/sha}"
+kubectl get --selector=app=batch-deployment deployments -o "jsonpath={.items[*].metadata.labels.hail\.is/sha}"
```
The labels are just the app name:
```
(hail) dking@wmb16-359 # kubectl get --selector=app=hail-ci-deployment deployments -o "jsonpath={.items[*].metadata.labels.hail\.is/sha}"
(hail) dking@wmb16-359 # kubectl get --selector=app=batch-deployment deployments -o "jsonpath={.items[*].metadata.labels.hail\.is/sha}"
(hail) dking@wmb16-359 # kubectl get --selector=app=batch deployments -o "jsonpath={.items[*].metadata.labels.hail\.is/sha}"
dd91f3447b3a
(hail) dking@wmb16-359 # kubectl get --selector=app=hail-ci deployments -o "jsonpath={.items[*].metadata.labels.hail\.is/sha}"
dd91f3447b3a
(hail) dking@wmb16-359 # 
```
```
(hail) dking@wmb16-359 # kubectl describe deployment batch-deployment
Name:                   batch-deployment
Namespace:              default
CreationTimestamp:      Mon, 27 Aug 2018 21:23:33 -0400
Labels:                 app=batch
                        hail.is/sha=dd91f3447b3a
Annotations:            deployment.kubernetes.io/revision=85
                        kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"apps/v1beta2","kind":"Deployment","metadata":{"annotations":{},"labels":{"hail.is/sha":"dd91f3447b3a"},"name":"batch-deployment","namesp...
Selector:               app=batch
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:  app=batch
```

cseed replied:
Oops, my bad.  I thought I saw some unnecessary redeploys and thought this explained it.  I'll back out.

danking commented on upload/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+    apt-get -y install python python-pip && \
+    rm -rf /var/lib/apt/lists/*
+
+RUN pip install -U \
```
`--no-cache-dir` to keep the image small

danking commented on upload/Makefile:
```diff
@@ -0,0 +1,16 @@
+	echo $(IMAGE) > upload-image
+
+deploy:
+	sed -e "s,@sha@,$SHA," \
```
where is SHA set?

danking commented on hail-ci-deploy.sh:
```diff
@@ -1,8 +1,7 @@
 #!/bin/bash
```
I don't see a `hail-ci-deploy.sh` for `upload`?

Also, we should probably have a `hail-ci-build.sh` that at least starts the server once to check the python syntax.

danking replied:
I guess there's the issue of the secrets. Maybe create some bogus secrets? `upload` does not appear to eagerly use the secrets.

cseed replied:
> I guess there's the issue of the secrets.

What's the issue?  I don't understand.

danking replied:
I'm pushing back on the fact that this has no tests at all, not even a test that it has correct python syntax. The issue with the secrets is that `upload` requires some secret information to run, but `hail-ci-build.sh` is executed in an environment that does not have those secrets.

cseed replied:
Can I follow up with another PR that runs the script? I need to rebuild the image with a new environment for upload.

cseed commented:
New test failure mode:

ERROR: (gcloud.beta.dataproc.clusters.create) Operation [projects/broad-ctsa/regions/global/operations/07164d0e-6c27-35d9-8132-9960b0db6d43] failed: Internal server error.

cseed commented:
New new failure mode:

AccessDeniedException: 403 hail-ci-0-1@broad-ctsa.iam.gserviceaccount.com does not have storage.objects.list access to hail-ci-test.

Failing here in test-ci.py:

```
        deploy_artifact = run(['gsutil', 'cat', f'gs://hail-ci-test/{second_target_sha}'], stdout=subprocess.PIPE)
        deploy_artifact = deploy_artifact.stdout.decode('utf-8').strip()
        assert f'commit {second_target_sha}' in deploy_artifact
```

I don't know who's authorizing gcloud for hail-ci-0-1.  Anyway, I gave hail-ci-0-1 permissions on hail-ci-test and am re-running.

danking commented:
Interesting, hail-ci-0-1 should have already had permissions to that bucket. I wonder if this is a symptom of some underlying issue.

cseed commented:
I agree.  I'll study up on testing this stuff.  Scorecard isn't tested, either.  A few thoughts:
 - I don't feel quite so bad having some of this untested (scorecard, etc.) while we get up to speed since they are internal tools (and not too complicated, unlike ci), but at the very least we need to test hl.upload_log() since that's the user facing bit.
 - It will get easier to run tests if we can deploy the service in a test namespace to mirror the production namespace.  I'll bump up the priority on looking into this.
 - We need authentication without oauth2 for the tests.  I'm at a total loss about how to automate testing of oauth2 login.  The internet has some thoughts: https://stackoverflow.com/questions/39180008/automated-api-testing-of-oauth2-openid-connect-protected-api, including using headless automation: https://medium.com/@vicusbass/api-testing-with-rest-assured-oauth2-flow-with-redirect-uri-ba48b5953823
 - Flask has a test fixture, so at least I can write local tests: http://flask.pocoo.org/docs/1.0/testing/
 - Created an issue to track these: https://github.com/hail-is/hail/issues/4539
